cypress/integration/9-room-status-bar/account-suspension.ts

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/// <reference types="cypress" />
+
+import { SynapseInstance } from "../../plugins/synapsedocker";
+
+describe("Room Status Bar", () => {
+    let synapse: SynapseInstance;
+    let roomId: string;
+
+    beforeEach(() => {
+        // Start a Synapse, create a user, create a room.
+        return cy.startSynapse("default").then(data => {
+            synapse = data;
+            return cy.initTestUser(synapse, "Alice");
+        }).then(_ => {
+            return cy.createRoom({});
+        }).then(data => {
+            roomId = data;
+            cy.visit("/#/room/" + roomId);
+        });
+    });
+
+    afterEach(() => {
+        //cy.stopSynapse(synapse);
+    });
+
+    const TEXT = "Hello, world";
+    it("shouldn't display an error message when there is no error", () => {
+        // User sends message
+        cy.get(".mx_RoomView_body .mx_BasicMessageComposer_input").type(`${TEXT}{enter}`);
+
+        // Wait for message to send
+        cy.get(".mx_RoomView_body .mx_EventTile").contains(".mx_EventTile[data-scroll-tokens]", TEXT);
+
+        // Give an error a little time to show up. It shouldn't.
+        cy.wait(1_000);
+        cy.get(".mx_RoomStatusBar_unsentMessages .mx_RoomStatusBar_unsentTitle").should('not.exist');
+    });
+
+    // Pretend to send a message but inject an error response.
+    function sendWithErrorResponse(text, response) {
+        cy.intercept("http://localhost:*/_matrix/client/r0/rooms/*/send/m.room.message/*", req => {
```

Comment:
> can we not just call a synapse API to suspend the user?

Not yet.

Reply:
Testing shouldn't be blocked by a frontend implementation, otherwise we'd get nothing done :p 

This sitting on a branch/here is fine while the rest of the feature is worked out, particularly considering this will be needing design review as well.

For the test here it's moderately important that we don't just intercept random endpoints and instead use the API provided by Synapse to do things: this means getting an admin API to suspend a user.

Reply:
> Testing shouldn't be blocked by a frontend implementation, otherwise we'd get nothing done :p

I actually meant user-testing. That can't happen without a frontend.

> For the test here it's moderately important that we don't just intercept random endpoints and instead use the API provided by Synapse to do things: this means getting an admin API to suspend a user.

I don't understand why it's important, but I'll take your word on it.

However, designing an API and getting it landed in Synapse before we can do any kind of user test sounds like something that will take us back several weeks, at best. The entire idea of prototyping it as a spam-check module is to be able to iterate quickly and without friction. I would very much prefer that.

Reply:
hmm, I'm not aware of a user testing plan for this to be honest 😬 

Because we're doing these tests at the end-to-end layer, we should be ensuring that it's, well, end-to-end 😇 For endpoints and such, it means performing actions as "server admins" to suspend/deactivate users and reacting to how the real server will tell us about it. If we just intercept the `/send` endpoint then we're no better off than a unit test, which doesn't help us test the application state.

A spam checking module is a good option, however suspension would also need to interact in a lot of other ways to prevent sustained abuse towards the server.

Reply:
> A spam checking module is a good option, however suspension would also need to interact in a lot of other ways to prevent sustained abuse towards the server.

Yes. The idea is to progressively experiment with this.

Reply:
@turt2live Would you be more comfortable with the idea of letting this go through if we made the following changes?

1. Open a MSC for an error code (and associated field) `M_BLOCKED_BY_SPAMCHECK` (plus a prefixed version), dedicated specifically to being sent by spam-checker.
2. Rewrite all the user testing code we have to issue `M_BLOCKED_BY_SPAMCHECK` (prefixed).
3. Rewrite the current PR to use `M_BLOCKED_BY_SPAMCHECK` (prefixed).

This would avoid the current conundrum of pretending-to-be-a-synapse-feature because it clarifies that this is not a synapse feature but rather a spam-checker feature.

Reply:
I don't think an MSC would be needed for such an error code as that'd be a bit too far on the implementation-specific side of things.

We would still need tests that prove the spam checker error code is showing up appropriately, which is still non-zero levels of effort.

Reply:
> We would still need tests that prove the spam checker error code is showing up appropriately, which is still non-zero levels of effort.

It feels to me that this is the responsibility of the spam-checker, rather than that of matrix-react-sdk. Do you concur?

Reply:
It's not a challenge to detect the error code, but we absolutely need to have the `if` statement tested because otherwise it will break. Our end-to-end tests are deliberately set up to *not* mock error responses as we feel it's important to test that the actual server implementation works. 

If we were interested in mocked errors, we'd have used Complement or similar to set up a mocked homeserver. We picked Synapse to ensure we're able to function with a proper server, though. 

We cannot land this without suitable tests, so to unblock this the tests need to be written. It is currently the only blocker for this PR.

Reply:
> We cannot land this without suitable tests, so to unblock this the tests need to be written. It is currently the only blocker for this PR.

I'm currently doing my best to find a way to write a test that matches your criteria without having to:

1. throw away several weeks worth of work by three people;
2. start several more weeks of work in a different direction that doesn't match our requirements.

Is there any chance you could find an alternative solution?

Reply:
> A plausible test is one which installs the spam checker module onto a dedicated synapse (since we can support running different flavours of Synapse already) and trigger that with a message of some kind.

Very well, what's the procedure to create a dedicated Synapse for use with cypress?

Reply:
We have a "consent" template already that I think you should be able to largely copy/paste. Visiting [#element-dev:matrix.org](https://matrix.to/#/#element-dev:matrix.org) would be better than asking here though, to spread the support load and get an answer more quickly.

Reply:
Thank you.